### ISSUE TITLE
`divmod` not returning correct types

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#10849]
**Describe the bug**
If creating overloads for `__divmod__()` that should be applied with `divmod`, the wrong types are returned when using `divmod`

**Code or Screenshots**
File `divmod.pyi` :
```python
from typing import overload

class LHS:
    @overload
    def __divmod__(self, other: RHS) -> tuple[RHS, RHS]: ...  # type: ignore[overload-overlap]
    @overload
    def __divmod__(self, other: object) -> tuple[LHS, LHS]: ...

class RHS:
    @overload
    def __divmod__(self, other: LHS) -> tuple[LHS, LHS]: ...  # type: ignore[overload-overlap]
    @overload
    def __divmod__(self, other: object) -> tuple[RHS, RHS]: ...
```

File `divmodtst.py` :
```python
from divmod import LHS, RHS

lhs = LHS()
rhs = RHS()

reveal_type(divmod(lhs, rhs))
reveal_type(lhs.__divmod__(rhs))
```

`pyright divmodtst.py` has this:
```text
divmodtst.py:1:6 - warning: Import "divmod" could not be resolved from source (reportMissingModuleSource)
divmodtst.py:6:13 - information: Type of "divmod(lhs, rhs)" is "tuple[RHS, RHS] | tuple[LHS, LHS]"
  c:\Code\Misc\pytypes\divmodtst.py:7:13 - information: Type of "lhs.__divmod__(rhs)" is "tuple[RHS, RHS]"
```

`mypy divmodtst.py` has this:
```text
divmodtst.py:6: note: Revealed type is "tuple[divmod.RHS, divmod.RHS]"
divmodtst.py:7: note: Revealed type is "tuple[divmod.RHS, divmod.RHS]"
```

Note that the results for `pyright` for calling `divmod` vs. `__divmod__()` are different.  I think they should be the same.

**VS Code extension or command-line**

pyright command linen version 1.1.404, python 3.11

**Comment by erictraut** [comment]
The behavior you're seeing is because of a change I made for [this issue](https://github.com/microsoft/pyright/issues/9835). I expressed concern at the time that pyright's behavior would deviate from mypy's in this case. Perhaps it would be best to revert it. cc @jorenham

**Comment by jorenham** [comment]
In https://github.com/microsoft/pyright/issues/9896#issuecomment-2691363462 I (failed to) explain that this can be solved by using the intersection of return types of matching overloads. That's because if multiple overloads match, then argument types are assignable to the intersection of matching input types signatures. Then, if you follow the `->`, you end up with the intersection of return types. The problem we're seeing here is because the union type is taken instead of the intersection type.
So I don't see the need for reverting this.

**Comment by erictraut** [comment]
@jorenham, I don't want to go through that debate again. I disagree with your analysis on that point. An intersection is not the right answer here. It sounds like you're in agreement that the current behavior is incorrect, so reverting it is appropriate. That will make pyright agree with mypy, as it previously did. Mypy's behavior here is also demonstrably incorrect (since it takes into account only the first overload).

@Dr-Irv, pyright's current behavior would work fine here if you eliminate the unsafe overload overlap (the error that you've suppressed in your code sample).

**Comment by Dr-Irv** [comment]
> Mypy's behavior here is also demonstrably incorrect (since it takes into account only the first overload).

I agree.  It's much easier to understand that a first matching overload is the one that gets chosen.

> @Dr-Irv, pyright's current behavior would work fine here if you eliminate the unsafe overload overlap (the error that you've suppressed in your code sample).

I made up a simple example, but what we have in `pandas-stubs` is going to require that the unsafe overload overlaps (at least for mypy)

**Comment by jorenham** [comment]
> I don't want to go through that debate again.

That's something we can agree on then.


> It sounds like you're in agreement that the current behavior is incorrect, so reverting it is appropriate.

That's right. I think the previous behavior was *more* incorrect though. 

Either way, I'll probably open a discourse thread about this at some point. I've managed to work around it in the codebases I maintain, so for me it's not such a pressing issue anymore. So do whatever you want I suppose. I'd just like to ask you to keep an eye on mypy_primer if you decide to revert it. It's safe to say that it took me several days to figure out how to work around workaround this, and I don't want to have to do that again. 

Also, I really appreciate the ping, especially given the earlier debate we had about this. So thanks for that.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.405
